serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,288 @@
+			// It's the same token.
+			return false, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+		}
+		c.config = config
+		// Continue to try again, but return the error too in case the caller would rather read it out.
+		return true, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	case 404:
+		return false, &ErrNotFound{debuggingInfo: sanitizedDebuggingInfo(req, reqBody, resp)}
+	case 500, 502, 503, 504:
+		// Could be transient.
+		return true, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	default:
+		// Unexpected.
+		return false, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	}
+
+	// We only arrive here with success.
+	// If we're not supposed to read out the body, we have nothing further
+	// to do here.
+	if ptrToReturnObj == nil {
+		return false, nil
+	}
+
+	// Attempt to read out the body into the given return object.
+	return false, json.NewDecoder(resp.Body).Decode(ptrToReturnObj)
+}
+
+type Pod struct {
+	Metadata *Metadata `json:"metadata,omitempty"`
+}
+
+type Metadata struct {
+	Name string `json:"name,omitempty"`
+
+	// This map will be nil if no "labels" key was provided.
+	// It will be populated but have a length of zero if the
+	// key was provided, but no values.
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+type PatchOperation int
+
+const (
+	// When adding support for more PatchOperations in the future,
+	// DO NOT alphebetize them because it will change the underlying
+	// int representing a user's intent. If that's stored anywhere,
+	// it will cause storage reads to map to the incorrect operation.
+	Unset PatchOperation = iota
+	Add
+	Replace
+)
+
+func Parse(s string) PatchOperation {
+	switch s {
+	case "add":
+		return Add
+	case "replace":
+		return Replace
+	default:
+		return Unset
+	}
+}
+
+func (p PatchOperation) String() string {
+	switch p {
+	case Unset:
+		// This is an invalid choice, and will be shown on a patch
+		// where the PatchOperation is unset. That's because ints
+		// default to 0, and Unset corresponds to 0.
+		return "unset"
+	case Add:
+		return "add"
+	case Replace:
+		return "replace"
+	default:
+		// Should never arrive here.
+		return ""
+	}
+}
+
+type Patch struct {
+	Operation PatchOperation
+	Path      string
+	Value     interface{}
+}
+
+type ErrNotFound struct {
+	debuggingInfo string
+}
+
+func (e *ErrNotFound) Error() string {
+	return e.debuggingInfo
+}
+
+// sanitizedDebuggingInfo converts an http response to a string without
+// including its headers to avoid leaking authorization headers.
+func sanitizedDebuggingInfo(req *http.Request, reqBody []byte, resp *http.Response) string {
+	respBody, _ := ioutil.ReadAll(resp.Body)
+	return fmt.Sprintf("req method: %s, req url: %s, req body: %s, resp statuscode: %d, resp respBody: %s", req.Method, req.URL, reqBody, resp.StatusCode, respBody)
```

Comment:
Request body could have sensitive information that is being logged verbatim.

Reply:
That's a good question, I'm not certain. Are you thinking of any particular sensitive information? @jasonodonnell , what do you think?

My take is that I think it should be safe to log response bodies, since the present calls are for Get and Patch for pods. The authentication token is in the header, so that won't be contained. I suppose that if the person were _explicitly_ writing unencrypted secrets in their pod json/yaml, that could be leaked, but they shouldn't do that in general, using [secrets](https://kubernetes.io/docs/concepts/configuration/secret/) instead.

Reply:
I think the risk with a Kube client logging like this is if we extended the client to support more of the Kube API, it's going to become harder to vet this.

Reply:
Stripped the request and response bodies from the output: https://github.com/hashicorp/vault/pull/8249/commits/662c272086c54a05515a44ec79fc3882c2900965